### DISCUSSION TITLE
GitHub OAuth2.0 use GithubApi in an Angular Page

### DISCUSSION THREAD

**Discussion by Timschmid** [community/community#46253] [Apps, API and Webhooks]
### Select Topic Area

Product Feedback

### Body

Hello,
I have an Angular Page, where I try to access The GitHub API via OAuth2.0.
I've already installed a GitHub OAuth2 App.

![image](https://user-images.githubusercontent.com/22547772/216551752-9fce48d3-be2c-420c-a384-cd4ae4d9d53c.png)

I use the library "angular-oauth2-oidc" in Angular, but I need the URL of the ISSUER. Thahts my config.

![image](https://user-images.githubusercontent.com/22547772/216552838-b41437e7-7765-4281-8006-c83853fbf298.png)

When I call the Login Method, I just get redirected to my localhost site.

![image](https://user-images.githubusercontent.com/22547772/216553269-94c365cd-d529-4b1d-aa65-c89e2a197705.png)

**Comment by jge162** [comment]
Is the first picture all the code? or you did not include imports on purpose in screenshot?

**Reply by jge162** [reply]
Here is a pseudo code example -> I have not tested it but tried to implement your idea. You should test for your use case 

```typescript
import { Component } from '@angular/core';
import { OAuthService } from 'angular-oauth2-oidc';

@Component({
  selector: 'app-root',
  template: `
    <button (click)="login()">Login</button>
  `
})
export class AppComponent {
  constructor(private oauthService: OAuthService) {
    this.oauthService.configure({
      issuer: 'https://github.com/login/oauth/authorize',
      strictDiscoveryDocumentValidation: false,
      redirectUri: window.location.origin,
      clientId: 'YOUR_GITHUB_APP_CLIENT_ID',
      scope: 'repo',
      responseType: 'code'
    });
  }

  async login() {
    try {
      await this.oauthService.initLoginFlowInPopup();
      const tokens = await this.oauthService.loadTokenInPopup();
      console.log('Access token:', tokens.access_token);
    } catch (error) {
      console.error('Error during login:', error);
    }
  }
}

**Reply by Timschmid** [reply]
this does not work

**Reply by jge162** [reply]
what was the error message?

**Reply by jge162** [reply]
if you could provide your code in code block versus pics would help alot, trying to piece it together in pics challenging here is updated
pseudo code ->
```typescript
import { AuthConfig, OAuthErrorEvent, OAuthService } from 'angular-oauth2-oidc';
import { Injectable } from '@angular/core';
import { HttpClientModule } from "@angular/common/http";
import { OAuthModule } from 'angular-oauth2-oidc';

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private oAuthConfig: AuthConfig = {
    issuer: 'https://github.com/login/oauth/authorize',
    strictDiscoveryDocumentValidation: false,
    redirectUri: window.location.origin,
    clientId: 'YOUR_GITHUB_APP_CLIENT_ID',
    scope: 'repo',
    responseType: 'code'
  };

  constructor(private oauthService: OAuthService) {
    this.oauthService.configure(this.oAuthConfig);
  }

  async login() {
    try {
      await this.oauthService.initLoginFlowInPopup();
      const tokens = await this.oauthService.loadTokenInPopup();
      console.log('Access token:', tokens.access_token);
    } catch (error) {
      console.error('Error during login:', error);
    }
  }
}
```

**Reply by jge162** [reply]
I did not see your snippets below from earlier, thanks

**Reply by Timschmid** [reply]
is it possible, that I have to load the DiscoveryDocument at first?

![image](https://user-images.githubusercontent.com/22547772/216569793-9d2a1a59-441c-478d-a2fd-00b155696a9e.png)

but then I get this error

**Reply by jge162** [reply]
yes you should 

try this -> 

```typescript
async login() {
    try {
      await this.oauthService.loadDiscoveryDocument();
      const isDiscoveryDocumentLoaded = await this.oauthService.setupAutomaticSilentRefresh();
      if (!isDiscoveryDocumentLoaded) {
        console.error('Error loading discovery document');
        return;
      }

      await this.oauthService.initLoginFlowInPopup();
      const tokens = await this.oauthService.loadTokenInPopup();
      console.log('Access token:', tokens.access_token);
    } catch (error) {
      console.error('Error during login:', error);
    }
  }
}
```

**Reply by Timschmid** [reply]
```
 async login(){      
    try {
      await this.oauthService.loadDiscoveryDocument();
      const isDiscoveryDocumentLoaded = await this.oauthService.setupAutomaticSilentRefresh();

      await this.oauthService.initLoginFlowInPopup();

    } catch (error) {
      console.error('Error during login:', error);
    }
  }

}
```
The if-clause doesn't work because it's a void, but if got error while request the discovery doc.


![image](https://user-images.githubusercontent.com/22547772/216573093-83f12099-bfe5-4ba6-a550-ccde8053d042.png)

**Reply by jge162** [reply]
You're correct, it was not a Boolean -> see updated use case (pseudo code). 

```typescript
async login() {
    try {
      await this.oauthService.loadDiscoveryDocument();
      this.oauthService.events.subscribe(e => {
        if (e.type === 'discovery_document_loaded') {
          console.log('Discovery document loaded');
        }
      });

      await this.oauthService.initLoginFlowInPopup();
      const tokens = await this.oauthService.loadTokenInPopup();
      console.log('Access token:', tokens.access_token);
    } catch (error) {
      console.error('Error during login:', error);
    }
  }
}
```

**Reply by Timschmid** [reply]
I Still get those Errors

**Reply by jge162** [reply]
hm, I do not know. the exact same ones?

**Reply by Timschmid** [reply]
yes

**Reply by jge162** [reply]
Verify the `issuer` URL
Verify the `client ID`
Check the browser's console log may contain more detailed information about the error. Check the console log for any error messages related to the `discovery document`

can you test authorization server? `curl https://github.com/login/oauth/authorize` to see if you get a response?

running out of ideas sorry mate

**Reply by Timschmid** [reply]
The Auth Server works correkt 

![image](https://user-images.githubusercontent.com/22547772/216580557-94d5089b-6b03-433b-9a7a-abaad18a5be5.png)

**Reply by jge162** [reply]
`httpErrorResponce { headers" HttpHeaders. status : 0, 'unkown error', url: 'Https://github.com/login/oauth/authorize/.well-kown/openid-configuration', ok: false. ..}` 

ok final hail Mary. make sure that the authorization server is allowing CORS and that the response includes the appropriate - `CORS headers`
ensure that the `issuer URL` in the `AuthConfig` is correct (which I believe they are). The URL should be the base URL of the authorization server's OAuth 2.0 configuration endpoint, which is typically `https://{authorization-server-domain}/.well-known/openid-configuration`

does the host server have the CORS headers?

**Reply by jge162** [reply]
these headers - >

```typescript
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept
```

**Reply by Timschmid** [reply]
Since I host the Page via GH-Pages I gues it does not have cors headers

**Comment by Timschmid** [comment]
No thats not alls of the Code ^^

![image](https://user-images.githubusercontent.com/22547772/216558157-82072b6b-b72a-4ac8-9f5c-20a12f8c9360.png)

here are the imports

**Comment by Timschmid** [comment]
![image](https://user-images.githubusercontent.com/22547772/216558320-6fde8691-cc79-45d1-a98b-43a82525d7d6.png)
